planner/core/task.go

```diff
@@ -1472,38 +1472,36 @@ func (p *PhysicalHashAgg) attach2Task(tasks ...task) task {
 		}
 	} else if mpp, ok := t.(*mppTask); ok {
 		// TODO : && t.count() > float64(p.SCtx().GetSessionVars().BroadcastJoinThresholdCount
-		if p.SCtx().GetSessionVars().AllowMPPExecution && p.tp == plancodec.TypeHashAgg {
-			if p.GetChildReqProps(0).PartitionTp == property.AnyType {
-				/// 1-phase agg: when the partition columns can be satisfied
+		if mpp.partTp == property.HashType {
+			if receiver, ok := mpp.p.(*PhysicalExchangeReceiver); !ok {
+				/// 1-phase agg: when the partition columns can be satisfied, where the plan does not need to enforce Exchange
 				/// only push down the original agg
 				p.self.SetChildren(mpp.p)
 				mpp.p = p.self
 				mpp.addCost(p.GetCost(inputRows, false))
 				return mpp
-			} else if mpp.partTp == property.HashType {
+			} else {
```

Comment:
I think the logic is reversed:

- If we find a exchanger, it means we need to do 2-phase agg, or don't do it if the gby column's cardinality is low.
- If we don't find a exchanger, it means the task **suffices** the partition requirements. We just need to do 1-phase agg.